File: mark-jenny/frontend/src/lib/api/workAgent.ts
import { api } from './client';

export interface WorkTask {
  id: string;
  type: string;
  description: string;
  status: string;
  created_at: string;
  started_at?: string;
  completed_at?: string;
  requires_approval: boolean;
  result?: Record<string, unknown>;
}

export interface WorkPattern {
  id: string;
  type: string;
  trigger: string;
  action: string;
  frequency: number;
  confidence: number;
  last_seen: string;
}

export interface WorkRecord {
  id: string;
  task_type: string;
  description: string;
  outcome: string;
  timestamp: string;
  duration: number;
}

export interface Prediction {
  need: string;
  confidence: number;
  suggestion: string;
  pattern_id?: string;
}

export const workAgentApi = {
  getTasks: () =>
    api.get<{ status: string; pending: WorkTask[]; active: WorkTask[]; completed: WorkTask[] }>('/work-agent/tasks'),

  createTask: (taskType: string, description: string, params?: Record<string, unknown>, requiresApproval?: boolean) =>
    api.post<{ status: string; task: WorkTask }>('/work-agent/tasks', {
      task_type: taskType,
      description,
      params: params || {},
      requires_approval: requiresApproval !== false,
    }),

  approveTask: (taskId: string) =>
    api.post<{ status: string }>(`/work-agent/tasks/${taskId}/approve`, {}),

  completeTask: (taskId: string, result?: Record<string, unknown>) =>
    api.post<{ status: string }>(`/work-agent/tasks/${taskId}/complete`, result || {}),

  failTask: (taskId: string, error?: string) =>
    api.post<{ status: string }>(`/work-agent/tasks/${taskId}/fail`, { error: error || 'Unknown error' }),

  fileAction: (action: string, params?: Record<string, unknown>) =>
    api.post<{ status: string; result: unknown }>('/work-agent/files', { action, params: params || {} }),

  draftEmail: (to: string, subject: string, context: string, tone?: string) =>
    api.post<{ status: string; result: unknown }>('/work-agent/email/draft', { to, subject, context, tone: tone || 'professional' }),

  readEmail: () =>
    api.get<{ status: string; result: unknown }>('/work-agent/email/read'),

  getPatterns: () =>
    api.get<{ status: string; patterns: WorkPattern[] }>('/work-agent/patterns'),

  learnPattern: (trigger: string, action: string) =>
    api.post<{ status: string }>(`/work-agent/patterns/learn?trigger=${encodeURIComponent(trigger)}&action=${encodeURIComponent(action)}`, {}),

  predict: () =>
    api.get<{ status: string; predictions: Prediction[] }>('/work-agent/predict'),

  getRecords: (limit?: number) =>
    api.get<{ status: string; records: WorkRecord[] }>(`/work-agent/records${limit ? `?limit=${limit}` : ''}`),

  getStats: () =>
    api.get<{ status: string; stats: Record<string, number> }>('/work-agent/stats'),
};

File: mark-jenny/backend/app/api/v1/endpoints/work_agent.py
"""Mark-Imti Work Agent API endpoints"""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional

from app.core.work_agent import WorkAgent, WorkTaskType

router = APIRouter(tags=["work-agent"])

_work_agent = None


def get_work_agent():
    global _work_agent
    if _work_agent is None:
        from app.core.model_caller import ModelCaller
        _work_agent = WorkAgent(model_caller=ModelCaller())
    return _work_agent


class CreateTaskRequest(BaseModel):
    task_type: str
    description: str
    params: dict = {}
    requires_approval: bool = True


class FileActionRequest(BaseModel):
    action: str
    params: dict = {}


class EmailRequest(BaseModel):
    to: str = ""
    subject: str = ""
    context: str = ""
    tone: str = "professional"


@router.get("/tasks")
async def list_tasks():
    agent = get_work_agent()
    pending = agent.get_pending_tasks()
    active = agent.get_active_tasks()
    completed = agent.get_completed_tasks()
    return {
        "status": "ok",
        "pending": [{"id": t.id, "type": t.task_type.value, "description": t.description,
                      "status": t.status.value, "created_at": t.created_at} for t in pending],
        "active": [{"id": t.id, "type": t.task_type.value, "description": t.description,
                     "status": t.status.value, "started_at": t.started_at} for t in active],
        "completed": [{"id": t.id, "type": t.task_type.value, "description": t.description,
                        "status": t.status.value, "completed_at": t.completed_at} for t in completed[:20]],
    }


@router.post("/tasks")
async def create_task(req: CreateTaskRequest):
    agent = get_work_agent()
    try:
        task_type = WorkTaskType(req.task_type)
    except ValueError:
        raise HTTPException(400, f"Invalid task type: {req.task_type}")
    task = agent.create_task(task_type, req.description, req.params, req.requires_approval)
    return {
        "status": "ok",
        "task": {"id": task.id, "type": task.task_type.value, "description": task.description,
                  "status": task.status.value, "requires_approval": task.requires_approval},
    }


@router.post("/tasks/{task_id}/approve")
async def approve_task(task_id: str):
    agent = get_work_agent()
    if agent.approve_task(task_id):
        return {"status": "ok", "message": "Task approved and started"}
    raise HTTPException(404, "Task not found")


@router.post("/tasks/{task_id}/complete")
async def complete_task(task_id: str, result: dict = {}):
    agent = get_work_agent()
    if agent.complete_task(task_id, result):
        return {"status": "ok", "message": "Task completed"}
    raise HTTPException(404, "Task not found")


@router.post("/tasks/{task_id}/fail")
async def fail_task(task_id: str, error: str = "Unknown error"):
    agent = get_work_agent()
    if agent.fail_task(task_id, error):
        return {"status": "ok", "message": "Task marked as failed"}
    raise HTTPException(404, "Task not found")


@router.post("/files")
async def file_action(req: FileActionRequest):
    agent = get_work_agent()
    result = await agent.execute_file_task(req.action, req.params)
    return {"status": "ok", "result": result}


@router.post("/email/draft")
async def draft_email(req: EmailRequest):
    agent = get_work_agent()
    result = await agent.draft_email({
        "to": req.to, "subject": req.subject,
        "context": req.context, "tone": req.tone,
    })
    return {"status": "ok", "result": result}


@router.get("/email/read")
async def read_email():
    agent = get_work_agent()
    result = await agent.read_email({})
    return {"status": "ok", "result": result}


@router.get("/patterns")
async def get_patterns():
    agent = get_work_agent()
    return {"status": "ok", "patterns": agent.get_learned_patterns()}


@router.post("/patterns/learn")
async def learn_pattern(trigger: str, action: str):
    agent = get_work_agent()
    agent.learn_pattern(trigger, action)
    return {"status": "ok", "message": "Pattern learned"}


@router.get("/predict")
async def predict_needs():
    agent = get_work_agent()
    predictions = agent.predict_needs()
    return {"status": "ok", "predictions": predictions}


@router.get("/records")
async def get_records(limit: int = 50):
    agent = get_work_agent()
    return {"status": "ok", "records": agent.get_records(limit)}


@router.get("/stats")
async def get_stats():
    agent = get_work_agent()
    return {"status": "ok", "stats": agent.get_stats()}
